### DISCUSSION TITLE
Backporting 3D features

### DISCUSSION THREAD

**Discussion by Helghast095** [godotengine/godot-proposals#6193] [3D]
It would be great to have some of the very basic 3D features backported from Godot 4 to Godot 3.5 (which is stable and most of people still use).

Those features would be:

1. Light source overlay:

![maxresdefault](https://user-images.githubusercontent.com/62181191/215269263-1c02f524-69e0-4c97-b42b-be0abba2667e.png)

This simple but great feature is already present in Godot 4 as "projector" property in light source options, it could be implemented in the same way to Godot 3.5.

![image (1)](https://user-images.githubusercontent.com/62181191/215269319-bdffb51e-ccee-43bc-930b-321aaf451e59.png)

It's a really standard and old feature that a lot of games use to create prettier looking light sources and it would be extremely useful for any kind of realistic or stylized 3D lighting.

2. Decals:

Decals are a node in Godot 4 

![decal](https://user-images.githubusercontent.com/62181191/215269512-5d1bef63-78f0-4678-98f2-2e03497d2331.png)

This is even more basic 3D feature which is incredibly useful for level design, various effects and in-game customization.
Decals can be used to create properly placed bullet holes, wounds, customizable stickers on cars, guns, characters, and also dirt, graffiti, cracks on walls and all various kinds of decorations for various objects.

![530-90_5e0ad40453051](https://user-images.githubusercontent.com/62181191/215270587-658a8cf7-4d13-4ac7-a815-74ef5a88bc32.jpg)
![echoco10](https://user-images.githubusercontent.com/62181191/215270588-38bb3391-c0a5-49ee-aa1a-aa0c89ee45ca.jpg)

There are two decal shaders available in asset library, however they are both quite useless as they have rather terrible performance, and they don't feature a proper cull mask like Godot 4 decals do which would be useful for separating object which receives decal, so the decal won't be casted into unwanted meshes nearby, for example this way characters won't be incorrectly included in decal projection on wall if they happen to be nearby.

3. SSAO:

SSAO is a very useful way to instantly add depth into scene, especially with dynamic lighting, it saves a ton of time for the developer and adds a lot of style to the game. Without it most sceneries look bland and flat, especially considering the fact that Godot 3.5 lightmapper doesn't support AO only baking so SSAO is the best option here.
 
![ssao_on_off_farming_simulator_22_ls22](https://user-images.githubusercontent.com/62181191/215270600-0b143dcb-5d4e-4821-913f-a6025106dd5a.jpg)

We all know how unoptimized SSAO is in Godot 3.5, it's pretty useless with even the most basic 3D graphics because of it's unacceptable performance and it also looks just bad, this issue was already mentioned here:
https://github.com/godotengine/godot/issues/17568
However SSAO in Godot 4 looks nice and is well optimized, so it would be great to have it's better version backported here to 3.5.

All of those features I mentioned are industry standard, decals are about 25 year old feature, light projectors were used in games from 2004 (Half Life 2) and even older, SSAO was invented in 2007.
Those features are commonly used in many games for good reasons and I can't emphasize enough how useful they would be here in Godot 3.5.
It's something that I was expecing when I started using Godot but to my surprise it wasn't here (in terms of SSAO it was just useless), since Godot 4 does it right maybe it's a good time to complete Godot 3.5 - the only stable version at the momment, and add basic features it's been missing for years.

I also want to show the kind of project I'm working on:
![prototype2](https://user-images.githubusercontent.com/62181191/215271554-74f5ae9c-1682-4663-9dfb-d3778c8d3ea8.jpg)

As you can see, shadows from the light cover are too sharp and look unnatural which is why the use of light overlay (light projector) would be a perfect way to fix it.
Decals would help in adding a significant amount of  detail into environment and proper (fast like in G4) SSAO would add a lot more depth into geometry itself.

It is often near impossible and extremely time consuming to rewrite our entire projects and start all over again, so please consider those improvements, I and many other people would certainly be grateful.

**Comment by Zireael07** [comment]
AFAIK both decals and SSAO depend on totally rewritten rendering, so only the projector texture could be feasily backported

Anyway Godot 4 Release Candidates are supposed to be within days, and the converter from 3 to 4 is working very well now, so there's no need to rewrite

**Reply by Helghast095** [reply]
I wish the converter worked well but it does not:
![crap1](https://user-images.githubusercontent.com/62181191/215273957-c5d60c4c-de22-4048-9e9f-ba822e293f23.jpg)
![crap2](https://user-images.githubusercontent.com/62181191/215273964-eb3bb4f9-9563-419d-8e34-99d68ce1da24.jpg)

(that's from beta 16, newest one)

**Reply by Zireael07** [reply]
oh. Yeah, IK is I think still a bit broken in the betas :( but my understanding is that @TokageItLab said that it will be fixed in time for the 4.0 stable

As for script errors, yeah there's still some things that you need to fix (File and Directory being some of the offenders) but the project loads up and works, it's not "rewrite from scratch"

**Reply by Helghast095** [reply]
I use standard animations, no IK here, so it's worrying that it corrupted all of them and it also seems to corrupt meshes normals, all lighting, environment settings don't seem to be properly converted either, skybox is gone too and the list goes on.
The thing is rewriting code isn't easy, it usually takes redesigning the code which is again very time consuming just to have few extra visual effects that should be there in 3.5 a long time ago.

In my opinion 3.5 should not be just pushed aside, it's fun to use it, so it could use some final additions.

**Reply by Dadaskis** [reply]
Nice to know that RCs are going to be released soon, but yea, gotta agree that project conversion is a bit broken, i hope it gets more improvement during RCs and i'd like to attempt to convert my project (which i did but without reports) to report issues, if that will help

**Reply by TokageItLab** [reply]
Unfortunately, there is no compatibility between Transform3DTrack in 3.x and Position/Rotation/ScaleTrack in 4.0. You will need to re-import the animations from glTF.

**Comment by Dadaskis** [comment]
I totally support the idea of backporting mentioned features as developer of first person shooter

**Comment by filipex2000** [comment]
I believe that just like the navigation systems, several other features introduced in Godot 4 should be backported to Godot 3.5
